docs/design/datamover.md

```diff
@@ -0,0 +1,220 @@
+
+## Goals
+* Create an extensible data mover solution
+* Supply a default data mover option 
+* Supply APIs for DataMover CRs (eg: VolumeSnapshotBackup, VolumeSnapshotRestore, DataMoverClass)
+* Supply a sample codebase for the Data Mover plugin and controller implementation
+
+
+## Non Goals
+* Maintain 3rd party data mover implementations
+* Adding a status watch controller to Velero
+
+## User stories
+
+Story 1: 
+As an application developer, I would like to save the CSI snaphots in a S3 bucket. 
+
+Story 2:
+As a cluster admin, I would like to be able to restore CSI snapshots if disaster happens.
+
+## Design & Implementation details
+
+This design supports adding the data mover feature to the Velero CSI plugin and facilitates integrating various vendor implemented data movers. 
+
+![DataMover CRD](../images/datamovercrd.png)
+
+Note: We will be supporting VolSync as the default data mover. 
+
+There will be two controllers - VolumeSnapshotBackup & VolumeSnapshotRestore. The VolumeSnapshotBackup Controller will be responsible for reconciling VolumeSnapshotBackup CR. Likewise, VolumeSnapshotRestore Controller will watch for VolumeSnapshotRestore CR. Both of these CRs will have a reference to a DataMoverClass. 
+
+`DataMoverClass` is a cluster scoped Custom Resource that will have details about the data mover. The specified mover will be registered in the system by creating the datamoverclass CR, addig a velero plugin that will create the appropriate resources for datamovement of a single datamoverclass and a controller that will reconcile the objects created by the plugin. The datamoverclass spec will also include a field (`selector`) to identify the PVCs that would be moved with the given data mover.
+
+
+```
+apiVersion: volumesnapshotmover.datamover.io/v1alpha1
+kind: DataMoverClass
+metadata:
+  annotations:
+    volumesnapshotmover.datamover.io/default: "true"
+  name: <name>
+spec:
+  mover: <VolSync>
+  selector: <tagname>
+
+```
+
+The above `DataMoverClass` name will be referenced in `VolumeSnapshotBackup` & `VolumeSnapshotRestore` CRs. This will help in selecting the data mover implementation during runtime. If the `DataMoverClass` name is not defined, then the default `DataMoverClass` will be used, which in this case will be `VolSync`
+
+### Data Mover Backup
+
+When a velero backup is created, it triggers the custom velero CSI plugin plugin (velero BackupItemAction plugin) to create the `VolumeSnapshotBackup` CR in the app namespace. The extended plugin looks up for the PVCs in the user namespace mentioned in the velero backup and creates a `VolumeSnapshotBackup` CR for every PVC in that namespace that is filtered by the `datamoverclass.spec.selector`.
+
+
+```
+apiVersion: volumesnapshotmover.datamover.io/v1alpha1
```

Comment:
`datamover.oadp.openshift.io/v1alpha1`

Reply:
I am trying to keep this generic as possible and not have references to oadp/openshift.  Will be easier to submit a link to this proposal. Its very rare that someone is going to dig into code while reviewing it.